### DISCUSSION TITLE
Dynamic construction from JSON

### DISCUSSION THREAD

**Discussion by AloisMahdal** [pydantic/pydantic#5534] [Question] [answered]
I'm trying to build a system where I can build tree of items from a corresponding data.  Each item is of same type (subclass of generic Item), and tree can be arbitrarily complex.
﻿
I have [a working example of where I'm now][g] in a GitHub gist; here's the most important part:

  [g]: https://gist.github.com/AloisMahdal/dcf47fb8bebb85b6f0ca1c30db03ce87
﻿
```python

    def _item_from_data(data: Dict):
        require_data_type(data, dict)
        require_data_key(data, 'item_type')
        if data['item_type'] == 'bar_i':
            require_data_key(data, 'data')
            require_data_type(data['data'], dict)
            return BarItem(
                data=data['data'],
            )
        if data['item_type'] == 'foo_i':
            # [..]
            return FooItem(
                name=data['name'],
                age=data['age'],
            )
        if data['item_type'] == 'multi_i':
            require_data_key(data, 'children')
            require_data_type(data['children'], list)
            return MultiItem(
                children=[_item_from_data(c) for c in data['children']],
            )
```
﻿
I'm sure I'm not using all the Pydantic magic.  Sure, I can construct any individual item "kwargs" style, but I haven't figured out how to make Pydantic **choose sub-class** based on the data, so I end up with a `if`-laden factory.   Also in cases like `multi_i`, I feel that the children construction should also be possiible to delegate to Pydantic.  This forces me to do lot of the validation "manually" by own functions (just to allow myself to use Python syntax to initiate each child manually, etc.)
﻿
What is the Pydatnic way?  Which part of the manual should I re-read?

**Comment by samuelcolvin** [comment]
Not sure I understand completely, but pydantic V2 has very good support for recursive generics, you might try that and let us know how you get on.

The docs for this aren't published yet, but there are examples in the tests, and we can help here if you have queries hopefully.

**Reply by AloisMahdal** [reply]
I'll try to clarify.

I have data like this:

    {
        'item_type': 'multi_i',
        'children': [
            {
                'item_type': 'foo_i',
                'name': 'Joe',
                'age': 32,
            },
            {
                'item_type': 'bar_i',
                'data': {'a': 1, 'b': 2},
            },
            {
                'item_type': 'multi_i',
                'children': [
                    {
                        'item_type': 'foo_i',
                        'name': 'Bobby',
                        'age': 2,
                    },
                    {
                        'item_type': 'foo_i',
                        'name': 'Jimmy',
                        'age': 5,
                    },
                ],
            },
            {
                'item_type': 'BAD_TYPE',
            },
        ]
    }

I have an `Item` class and for every possible `item_type` I have a corresponding sub-class which takes the rest of keys as constructor arguments. For example, there's `class FooItem` (`foo_i`) with `name: str` and `age: int`, a `class MultiItem` (`multi_i`) with `children: Item`, etc.

What I want is to call some construction mechanism that will traverse the whole JSON tree such that when it sees a dictionary with `item_type` it will select a corresponding class and call it with the rest of the keys, eventually translating the data to a tree like this:

    MultiItem(children=[
        FooItem(name='Joe', age=32),
        BarItem(data={'a': 1, 'b': 2}),
        MultiItem(children=[
            FooItem(name='Bobby', age=2),
            FooItem(name='Jimmy', age=5)
        ]),
        InvalidItem(error='unknown condition type', error_details={'item_type': 'BAD_TYPE'}, data={'item_type': 'BAD_TYPE'})
    ])

The gist code I've linked in the OP does what I want but "manually" with a recursive factory function, which forces me to do some of the validation manually as well.

I was wondering if there is a more elegant way, perhaps using Pydantic features.

**Comment by chrisK824** [comment, marked as answer]
Kinda late to the party but @AloisMahdal I think what you need is the discriminators feature ; basically pydantic feature that can discriminate objects to be validated into different validator classes based on a field. https://docs.pydantic.dev/latest/usage/types/#discriminated-unions-aka-tagged-unions

so according to docs, something like that in your case
```
class Foo(BaseModel):
    item_type: Literal['foo_i']
    foo_specific_field: float


class Bar(BaseModel):
    item_type: Literal['bar_i']
    bar_specific_field: bool


class Item(BaseModel):
    item: Union[Foo, Bar] = Field(..., discriminator='item_type')
```

**Reply by chrisK824** [reply]
@AloisMahdal Have you tried the above or found another solution for this?

**Reply by AloisMahdal** [reply]
You hit the nail on the head, @chrisK824; it looks like exactly what I was looking for.

Unfortunately in our team we're using single environment which is still locked on 1.8.2 which does not contain this feature yet.

In the meantime I've also changed the architecture quite a bit. I'll describe it here in case someone has the same request for the same reasons.

My original intent was for the system to be extensible by adding `Item` subclasses.  The items actually represented actions and the real difference between them was single method and set of parameters it would require.

The current architecture is different:

 1. I've built a dedicated Factory class which builds the tree according to `item_type`s in such way that most validation is still left to Pydantic.

 2. The tree is now homogenic: all items are `Item`s.  I've also dropped the whole idea of `InvalidItem` -- it's just *ValidationError*, raised as it should be.

 3. They differ only in `callback`, which is a callable registered and configured by decorator.

    So `item_type` discerns which callback it is, and every callback has attribute of `params_cls` -- this is a (pretty generic) class that the callback must accept.  It also specifies whether it needs children or not, and few more domain-specific details.

    So that people can easily define the callbacks as functions and use decorator to register it as well as bind it to given `params_cls` and `'item_type'` in the JSON.

 4. I've learned how to generate JSON schema myself and do so for this `Item` class, since most of the schema is actually dictated by callbacks.

So the data now looks more like this:

```python
{
    'item_type': 'multi_i',
    'children': [
        {
            'item_type': 'foo_i',
            'params': {
                'name': 'Joe',
                'age': 32,
            },
        },
        {
            'item_type': 'bar_i',
            'params': {
                'data': {'a': 1, 'b': 2},
            }
        },
        {
            'item_type': 'multi_i',
            'children': [
                {
                    'item_type': 'foo_i',
                    'params': {
                        'name': 'Bobby',
                        'age': 2,
                    },
                },
                {
                    'item_type': 'foo_i',
                    'params': {
                        'name': 'Jimmy',
                        'age': 5,
                    }
                },
            ],
        },
        {
            'item_type': 'BAD_TYPE',
        },
    ]
}
```

And there's only `Item` class, but people can extend the system by adding callbacks:

```python
class Person(BaseModel):
    name: str
    age: int

class Data(BaseModel):
    data: dict[str, int]

@item(need_children=True)
def multi_i(children: list[Item]):
    for c in children:
        ...

@item(params_cls=Person)
def foo_i(params: Person):
    print(f"foo_i():params.name={params.name}")
    print(f"foo_i():params.age={params.age}")
    ...

@item(params_cls=Data)
def bar_i(params: Data):
    print(f"bar_i():params.data.keys()={params.data.keys()}")
    ...
```

(the concept of 'children' is handled with only a bit more complexity).

I think it's for the better, as it turns out there are also some requirements on what item can be selected which are completely domain specific so they are better handled explicitly by local factory anyway.

It's still possible to do a lot with Pydantic (`param_cls` and its sub-types which are open to extension).

I've also replaced inheritance with composition, which is good.

Also defining callbacks with decorators is way easier, less error-prone and more fun.
